keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1066,8 +1076,361 @@ the remaining requests in that queue start getting faster service.  In
+
+#### Determining the width
+
+While one can imagine arbitrarily sophisticated algorithms for it (including
+exposing defining the width of requests via FlowSchema API), we want to start
+with something relatively simple to first get operational experience with it
+before investing into sophisticated algorithms or exposing a knob to users.
+
+In order to determine the function that will be approximating the `width` of
+a request, we should first estimate how expensive a particular request is.
+And we need to think about both dimensions that we're trying to protect from
+overloading (CPU and RAM) and how many concurrency units a request can actually
+consume.
+
+Let's start with CPU.  The total cost of processing a LIST request should be
+proportional to the number of processed objects.  However, given that in
+practice processing a single request isn't parallelized (and the fact that
+we generally scale the number of total concurrency units linearly with amount
+of available resources), a single request should consume no more than A
+concurrency units.  Fortunately that all compiles together because the
+`processing latency` of the LIST request is actually proportional to the
+number of processed objects, so the cost of the request (defined above as
+`<width> x <processing latency>` really is proportaional to the number of
+processed objects as expected.
+
+For RAM the situation is actually different.  In order to process a LIST
+request we actually store all objects that we process in memory. Given that
+memory is uncompressable resource, we effectively need to reserve all that
+memory for the whole time of processing that request.  That suggests that
+the `width` for the request from the RAM perspective should be proportional
+to the number of processed items.
+
+So what we get is that:
+```
+  width_cpu(N) = min(A, B * N)
+  width_ram(N) = C * N
+```
+where N is the number of items a given LIST request is processing.
+
+The question is how to combine them to a single number.  While the main goal
+is to stay on the safe side and protect from the overload, we also want to
+maxiumize the utilization of the available concurrency units.
+Fortunately, when we normalize CPU and RAM to percentage of available capacity,
+it appears that almost all requests are much more cpu-intensive.  Assuming
+4GB:1CPU ratio and 10kB average object and the fact that processing larger
+number of objects can utilize exactly 1 core, that means that we need to
+process 400.000 objects to make the memory cost higher.
+This means, that we can afford the potential minor efficiency that extremely
+large requests would cause and just approximate it by protecting every resource
+independently, which translates to the following function:
+```
+  width(n) = max(min(A, B * N), C * N)
+```
+We're going to better tune the function based on experiments, but we will
+simply start with
+```
+width_approx(n) = max(A, ceil(N / B))
```

Comment:
I do not see how this formula follows from the previous one (which says it is for the same quantity).

I think a reasonable first design is a simple linear relationship (as is shown here), with rounding up.

Why have a cap?  A modest cap would enable us to use that cap as a floor on the concurrency limit for each priority level.  But I am not sure we would want a cap that is low enough for that.  Let's see if we can do without a cap.

Reply:
> I do not see how this formula follows from the previous one (which says it is for the same quantity).

Naming collision - it's not the same B... Will fix

> Why have a cap?

Yeah - this doesn't make any sense.
What I really meant here was based on the cpu:memory ratio, I wanted to suggest that we simply ignore memory and say that it will be simply:
```
min(A, B*N)
```

which we will approximate by min(A, ceil(N/D))  [where D = 1/B]

Does that make sense now?
[I will update]